solutions/gold/usaco-925.mdx

```diff
@@ -0,0 +1,127 @@
+---
+id: usaco-925
+source: USACO Platinum 2019 February
+title: Mooriokart
+author: Justin Ji
+---
+
+[Official Editorial (C++)](https://usaco.org/current/data/sol_mooriokart_platinum_feb19.html)
+
+<Spoiler title = "Hint 1">
+
+The graph is a forest, and we are only allowed to pick one path from each tree
+to reach our total. How can we write a knapsack DP to find our answer?
+
+</Spoiler>
+
+<Spoiler title = "Hint 2">
+
+The issue with directly writing a knapsack is that there is no bound on
+how long we can make our path. However, note how all paths of length $\ge Y$ can
+be compressed because they already reach the threshold.
+
+</Spoiler>
+
+<Spoiler title = "Explanation">
+
+We can first ignore the part about ordering the trees, and think of the problem
+from a knapsack DP perspective. Our goal is to choose a path from each tree that
+sums up to something $\ge Y$.
+
+For each tree, we can compress all the paths into $(length, frequency)$ pairs.
+Then, we can knapsack as usual. However, note that because we are compressing
+all paths of length $\ge Y$, we need to explicitly store the sum of the paths
+when writing our knapsack.
+
+Our DP state would be keep track of both the sum of the path lengths and
+the number of such paths, for each possible length of our path.
+
+Let $k$ be the number of trees in our forest. To account for the possible ways
+of ordering our paths, we need to multiply our initial sum of paths by
+$2^{k} * k!$. However, cyclic rotations and also reversed paths are considered
+to be the same, so we only multiply by $2^{k - 1} * (k - 1)!$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(NY^{\frac{3}{2}})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+constexpr int MOD = 1e9 + 7;
+int main() {
+	freopen("mooriokart.in", "r", stdin);
+	freopen("mooriokart.out", "w", stdout);
+	int n, m, x, y;
+	cin >> n >> m >> x >> y;
+	vector<vector<array<int, 2>>> adj(n);
+	for (int i = 0; i < m; i++) {
+		int u, v, d;
+		cin >> u >> v >> d;
+		adj[--u].push_back({--v, d});
+		adj[v].push_back({u, d});
+	}
+
+	vector<bool> vis(n);
+	vector<int> comp;
+	function<void(int, int)> dfs1 = [&](int u, int p) {
+		vis[u] = true;
+		comp.push_back(u);
+		for (auto [v, w] : adj[u]) {
+			if (v == p) { continue; }
+			dfs1(v, u);
+		}
+	};
+
+	map<int, array<int, 2>> occ;
+	function<void(int, int, int, int)> dfs2 = [&](int u, int p, int d, int s) {
+		for (auto [v, w] : adj[u]) {
+			if (v == p) { continue; }
+			if (v < s) {
+				auto &[sum, freq] = occ[min(y, d + w)];
+				sum = (sum + d + w) % MOD;
+				freq++;
+			}
+			dfs2(v, u, d + w, s);
+		}
+	};
+
+	vector<array<ll, 2>> dp(y + 1);
+	// dp[path_len] = {len_sum, num_paths}
```

Comment:
why do you put the comment below the thing

Reply:
oops - we're supposed to put it above the thing right? at least I think that's what u do lol